airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/streams.py

```diff
@@ -0,0 +1,422 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+import types
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+from urllib.parse import parse_qsl, urlparse
+
+import pytz
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+
+class SourceZendeskSupportStream(HttpStream, ABC):
+    """"Basic Zendesk class"""
+
+    primary_key = "id"
+
+    page_size = 100
+    created_at_field = "created_at"
+    updated_at_field = "updated_at"
+
+    def __init__(self, subdomain: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # add the custom value for generation of a zendesk domain
+        self._subdomain = subdomain
+
+    @property
+    def url_base(self) -> str:
+        return f"https://{self._subdomain}.zendesk.com/api/v2/"
+
+    @staticmethod
+    def _parse_next_page_number(response: requests.Response) -> Optional[int]:
+        """Parses a response and tries to find next page number"""
+        next_page = response.json()["next_page"]
+        # TODO test page
+        # next_page = """https://foo.zendesk.com/api/v2/search.json?page=2"""
+        if next_page:
+            return dict(parse_qsl(urlparse(next_page).query)).get("page")
+        return None
+
+    @staticmethod
+    def str2datetime(s):
+        """convert string to datetime object"""
+        if not s:
+            return None
+        return datetime.strptime(s, DATETIME_FORMAT)
+
+    @staticmethod
+    def datetime2str(dt):
+        """convert datetime object to string"""
+        return datetime.strftime(dt.replace(tzinfo=pytz.UTC), DATETIME_FORMAT)
+
+
+class UserSettingsStream(SourceZendeskSupportStream):
+    """Stream for checking of a request token and permissions"""
+
+    def path(self, *args, **kwargs) -> str:
+        return "account/settings.json"
+
+    def next_page_token(self, *args, **kwargs) -> Optional[Mapping[str, Any]]:
+        # this data without listing
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """returns data from API"""
+        yield from [response.json().get("settings") or {}]
+
+    def get_settings(self) -> Tuple[Mapping[str, Any], Union[str, None]]:
+        for resp in self.read_records(SyncMode.full_refresh):
+            return resp, None
+        return None, "not found settings"
+
+
+class IncrementalBasicSearchStream(SourceZendeskSupportStream, ABC):
```

Comment:
the search endpoint [returns a maximum of 1000 items per query](https://developer.zendesk.com/api-reference/ticketing/ticket-management/search/#results-limit). How can we make sure the relevant resources (users/groups/organizations/tickets) are properly loaded?

Reply:
I see your concern. I will add additional validations.

Reply:
I've updated this logic. Zendesk provides more comfortable approach for this
https://developer.zendesk.com/api-reference/ticketing/ticket-management/incremental_exports/